Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift

```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
                 switch(result) {
                 case .success(let output):
                     if let identityPools = output.identityPools {
-                        for pool in identityPools {
-                            if pool.identityPoolName == name,
-                               let poolId = pool.identityPoolId {
-                                completion(poolId)
-                                break
-                            }
+                        let foundID = identityPools.filter {
+                            $0.identityPoolName == name
+                        }
+                        .map {
+                            $0.identityPoolId
+                        }
+                        .first ?? nil
+                        
+                        if let foundID = foundID {
+                            poolID = foundID
                         }
                     }
                     
                     token = output.nextToken
                 case .failure(let error):
-                    completion(nil)
                     print("ERROR: ", dump(error, name: "Error scanning identity pools"))
+                    token = nil
+                    poolID = nil
                 }
             }
-        } while token != nil
+        } while token != nil && poolID == nil
         
-        completion(nil)
+        completion(poolID)
     }
     
     /// Returns the ID of the identity pool with the specified name.
     /// - Parameters:
     ///   - name: The name of the identity pool whose ID should be returned
     /// - Returns: A string containing the ID of the specified identity pool or `nil` on error or if not found
-    func getIdentityPoolID(name: String,
+    func getIdentityPoolID(name: String, createIfMissing: Bool = false,
                            completion: @escaping (String?) -> Void) {
         // See if the pool already exists
         
-        self.getIdentityPoolID(name: name) { poolId in
-            if let poolId = poolId {
-                completion(poolId)
+        self.getIdentityPoolID(name: name) { foundID in
+            if let foundID = foundID {
+                completion(foundID)
+            } else if createIfMissing {
+                self.createIdentityPool(name: name, completion: completion)
             } else {
-                self.createIdentityPool(name: name) { poolId in
-                    completion(poolId)
-                }
+                completion(nil)
             }
         }
-    }
+}
     
     /// Create a new identity pool, returning its ID.
     /// - Parameters:
     ///     - name: The name to give the new identity pool
     /// - Returns: A string containing the newly created pool's ID, or `nil` if an error occurred
     func createIdentityPool(name: String, completion: @escaping (String?) -> Void) {
-        
         let cognitoInputCall = CreateIdentityPoolInput(
             developerProviderName: "com.exampleco.CognitoIdentityDemo",
```

Comment:
Seems like the `developerProviderName` should be something injectable to this class, rather than hardcoded in this function.

Reply:
Yeah, that was intentional here though I see your point. Unless there's a strong objection I will revisit this after re:Invent.